File: zone_mapper.py
"""
Zone Mapper Module for Act 2 (Indian Education Machine)
Maps (x, y) offset from center into 8 discrete college-life directional zones
with center dead-zone filtering.
"""

import math

# Directional Zones with metadata
ZONES = [
    {
        "id": "assignments",
        "name": "Assignments",
        "direction": "NORTH",
        "angle_center": 270,  # Up
        "color": (56, 189, 248),  # Sky Blue
        "hex": "#38bdf8",
        "icon": "[ASSIGN]",
        "slang": "Copying senior's assignment at 2 AM",
        "flavor": "+Assignments | -Sleep"
    },
    {
        "id": "hackathon",
        "name": "Hackathon",
        "direction": "NORTH-EAST",
        "angle_center": 315,  # Up-Right
        "color": (250, 204, 21),  # Amber Yellow
        "hex": "#facc15",
        "icon": "[HACK]",
        "slang": "36-hr Red Bull sprint with ChatGPT",
        "flavor": "+Hackathon | -Friends, -Sleep"
    },
    {
        "id": "placement",
        "name": "Placement Prep",
        "direction": "EAST",
        "angle_center": 0,  # Right
        "color": (16, 185, 129),  # Emerald Green
        "hex": "#10b981",
        "icon": "[PLACEMT]",
        "slang": "Grinding LeetCode & Aptitude tests",
        "flavor": "+Placement | -Club Activities"
    },
    {
        "id": "club",
        "name": "Club Activities",
        "direction": "SOUTH-EAST",
        "angle_center": 45,  # Down-Right
        "color": (236, 72, 153),  # Pink/Magenta
        "hex": "#ec4899",
        "icon": "[CLUB]",
        "slang": "College fest coordinator drama",
        "flavor": "+Club | -Assignments"
    },
    {
        "id": "sleep",
        "name": "Sleep",
        "direction": "SOUTH",
        "angle_center": 90,  # Down
        "color": (129, 140, 248),  # Indigo
        "hex": "#818cf8",
        "icon": "[SLEEP]",
        "slang": "Passing out on hostel bed",
        "flavor": "+Sleep | -Placement"
    },
    {
        "id": "friends",
        "name": "Friends & Chai",
        "direction": "SOUTH-WEST",
        "angle_center": 135,  # Down-Left
        "color": (251, 146, 60),  # Orange
        "hex": "#fb923c",
        "icon": "[CHAI]",
        "slang": "Canteen Sulaimani & life debate",
        "flavor": "+Friends | -Record"
    },
    {
        "id": "mental_health",
        "name": "Panic / Cope",
        "direction": "WEST",
        "angle_center": 180,  # Left
        "color": (248, 113, 113),  # Red
        "hex": "#f87171",
        "icon": "[COPE]",
        "slang": "Staring at ceiling questioning life",
        "flavor": "Trying to stop mental health decay"
    },
    {
        "id": "record",
        "name": "Lab Record",
        "direction": "NORTH-WEST",
        "angle_center": 225,  # Up-Left
        "color": (6, 182, 212),  # Cyan
        "hex": "#06b6d4",
        "icon": "[RECORD]",
        "slang": "Drawing circuit diagrams in ink",
        "flavor": "+Record | -Sleep, -Friends"
    }
]


class ZoneMapper:
    def __init__(self, dead_zone_radius=0.18):
        self.dead_zone_radius = dead_zone_radius  # Radius in normalized [-1, 1] units
        self.zones = ZONES
        self.num_zones = len(ZONES)
        self.sector_angle = 360.0 / self.num_zones  # 45 degrees per sector

    def map_position(self, norm_x, norm_y):
        """
        Takes normalized (x, y) coordinates where (0, 0) is screen center,
        x in [-1, 1], y in [-1, 1] (+y is down).
        
        Returns:
            dict: {
                'is_dead_zone': bool,
                'distance': float,
                'angle_deg': float,
                'zone_index': int (or None if dead zone),
                'zone': dict (or None if dead zone)
            }
        """
        distance = math.sqrt(norm_x * norm_x + norm_y * norm_y)
        
        if distance < self.dead_zone_radius:
            return {
                "is_dead_zone": True,
                "distance": distance,
                "angle_deg": 0.0,
                "zone_index": None,
                "zone": None
            }
            
        # atan2 gives angle from positive x-axis (Right = 0 rad, Down = +pi/2, Up = -pi/2)
        angle_rad = math.atan2(norm_y, norm_x)
        angle_deg = math.degrees(angle_rad)
        if angle_deg < 0:
            angle_deg += 360.0
            
        # Each sector spans [center - 22.5, center + 22.5]
        # Sector 0 (Assignments / North) is centered at 270 deg
        # Sector 2 (Placement / East) is centered at 0 deg
        # Offset angle by 22.5 deg to align sector boundaries cleanly
        shifted_angle = (angle_deg + 22.5) % 360.0
        sector_idx = int(shifted_angle // 45.0)
        
        # Mapping sector_idx (where 0 is around 0° East) to our ZONES list:
        # 0 -> Placement (East) -> index 2
        # 1 -> Club (South-East) -> index 3
        # 2 -> Sleep (South) -> index 4
        # 3 -> Friends (South-West) -> index 5
        # 4 -> Cope (West) -> index 6
        # 5 -> Record (North-West) -> index 7
        # 6 -> Assignments (North) -> index 0
        # 7 -> Hackathon (North-East) -> index 1
        zone_map_table = [2, 3, 4, 5, 6, 7, 0, 1]
        active_zone_idx = zone_map_table[sector_idx]
        
        return {
            "is_dead_zone": False,
            "distance": min(1.0, distance),
            "angle_deg": angle_deg,
            "zone_index": active_zone_idx,
            "zone": self.zones[active_zone_idx]
        }

File: js/zone-mapper.js
/**
 * Zone Mapper Module for Act 2 (Indian Education Machine)
 * 4-Directional System:
 * ⬆️ UP    -> Academics    (Arts & Sports -20%)
 * ➡️ RIGHT -> Hackathons   (Academics -20%)
 * ⬇️ DOWN  -> Skill Dev    (Hackathons -20%)
 * ⬅️ LEFT  -> Arts & Sports (Skill Dev -20%)
 *
 * Supports both Browser (window.ZoneMapper) and Node.js (CommonJS).
 */

(function (root, factory) {
  if (typeof module === "object" && module.exports) {
    module.exports = factory();
  } else {
    const exportsObj = factory();
    root.ZONES = exportsObj.ZONES;
    root.ZONE_BY_ID = exportsObj.ZONE_BY_ID;
    root.ZoneMapper = exportsObj.ZoneMapper;
  }
})(typeof self !== "undefined" ? self : this, function () {
  const ZONES = [
    {
      id: "academics",
      name: "Academics",
      direction: "UP",
      arrow: "⬆️",
      angle_center: 270, // Up (Negative Y)
      color: [56, 189, 248],
      hex: "#38bdf8",
      decreases: "arts_sports",
      decreases_name: "Arts & Sports",
      penalty_text: "Arts & Sports (-20%)",
      slang: "Grinding KTU modules & exams"
    },
    {
      id: "hackathons",
      name: "Hackathons",
      direction: "RIGHT",
      arrow: "➡️",
      angle_center: 0, // Right (Positive X)
      color: [250, 204, 21],
      hex: "#facc15",
      decreases: "academics",
      decreases_name: "Academics",
      penalty_text: "Academics (-20%)",
      slang: "36-hr Red Bull code sprint"
    },
    {
      id: "skill_dev",
      name: "Skill Dev",
      direction: "DOWN",
      arrow: "⬇️",
      angle_center: 90, // Down (Positive Y)
      color: [16, 185, 129],
      hex: "#10b981",
      decreases: "hackathons",
      decreases_name: "Hackathons",
      penalty_text: "Hackathons (-20%)",
      slang: "Building portfolio & projects"
    },
    {
      id: "arts_sports",
      name: "Arts & Sports",
      direction: "LEFT",
      arrow: "⬅️",
      angle_center: 180, // Left (Negative X)
      color: [244, 114, 182],
      hex: "#f472b6",
      decreases: "skill_dev",
      decreases_name: "Skill Dev",
      penalty_text: "Skill Dev (-20%)",
      slang: "College fest, music & sports"
    }
  ];

  const ZONE_BY_ID = {};
  ZONES.forEach(z => {
    ZONE_BY_ID[z.id] = z;
  });

  class ZoneMapper {
    constructor(deadZoneRadius = 0.18) {
      this.dead_zone_radius = deadZoneRadius;
    }

    /**
     * Maps normalized coordinates (normX in [-1, 1], normY in [-1, 1] from center)
     * to one of the 4 directional zones.
     */
    mapPosition(normX, normY) {
      const distance = Math.sqrt(normX * normX + normY * normY);

      // Center dead zone: No zone active
      if (distance < this.dead_zone_radius) {
        return {
          zone: null,
          is_dead_zone: true,
          angle_deg: 0.0,
          distance: distance,
          direction: "CENTER"
        };
      }

      // Calculate angle in degrees [0, 360)
      // Note: In screen coordinates, positive Y is DOWN, negative Y is UP.
      const angleRad = Math.atan2(normY, normX);
      let angleDeg = (angleRad * 180) / Math.PI;
      if (angleDeg < 0) {
        angleDeg += 360.0;
      }

      // 4 Quadrants (45 degree boundaries):
      // 315° - 45°  (or >315 or <=45) -> RIGHT (Hackathons)
      // 45°  - 135°                   -> DOWN  (Skill Dev)
      // 135° - 225°                   -> LEFT  (Arts & Sports)
      // 225° - 315°                   -> UP    (Academics)
      let activeZone;
      if (angleDeg > 315.0 || angleDeg <= 45.0) {
        activeZone = ZONE_BY_ID["hackathons"];
      } else if (angleDeg > 45.0 && angleDeg <= 135.0) {
        activeZone = ZONE_BY_ID["skill_dev"];
      } else if (angleDeg > 135.0 && angleDeg <= 225.0) {
        activeZone = ZONE_BY_ID["arts_sports"];
      } else {
        activeZone = ZONE_BY_ID["academics"];
      }

      return {
        zone: activeZone,
        is_dead_zone: false,
        angle_deg: angleDeg,
        distance: Math.min(1.0, distance),
        direction: activeZone.direction
      };
    }

    // Alias for Python snake_case method compatibility
    map_position(normX, normY) {
      return this.mapPosition(normX, normY);
    }
  }

  return {
    ZONES,
    ZONE_BY_ID,
    ZoneMapper
  };
});
